File: frontend/src/services/dashboardService.ts
import api from './api';
import type { DashboardStats } from '../types';

const emptyStats: DashboardStats = {
  totalAssets: 0,
  availableAssets: 0,
  inUseAssets: 0,
  brokenAssets: 0,
  totalUsers: 0,
  adminUsers: 0,
  staffUsers: 0,
};

function normalizeStats(raw: unknown): DashboardStats {
  if (!raw || typeof raw !== 'object') {
    return { ...emptyStats };
  }
  const o = raw as Record<string, unknown>;
  const n = (v: unknown) => (typeof v === 'number' && !Number.isNaN(v) ? v : 0);
  return {
    totalAssets: n(o.totalAssets),
    availableAssets: n(o.availableAssets),
    inUseAssets: n(o.inUseAssets),
    brokenAssets: n(o.brokenAssets),
    totalUsers: n(o.totalUsers),
    adminUsers: n(o.adminUsers),
    staffUsers: n(o.staffUsers),
  };
}

export const dashboardService = {
  getStats: async (): Promise<DashboardStats> => {
    const response = await api.get<{ data?: DashboardStats }>('/dashboard/stats');
    return normalizeStats(response.data?.data);
  },
};

File: backend/src/main/java/com/lttn/quanlytaisan/service/DashboardService.java
package com.lttn.quanlytaisan.service;

import com.lttn.quanlytaisan.model.AssetStatus;
import com.lttn.quanlytaisan.repository.AssetRepository;
import com.lttn.quanlytaisan.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class DashboardService {

    private final AssetRepository assetRepository;
    private final UserRepository userRepository;

    public Map<String, Object> getDashboardStats() {
        log.debug("Fetching dashboard statistics");

        Map<String, Object> stats = new HashMap<>();

        stats.put("totalAssets", assetRepository.count());
        stats.put("availableAssets", assetRepository.countByStatus(AssetStatus.AVAILABLE));
        stats.put("inUseAssets", assetRepository.countByStatus(AssetStatus.IN_USE));
        stats.put("brokenAssets", assetRepository.countByStatus(AssetStatus.BROKEN));
        stats.put("totalUsers", userRepository.count());
        stats.put("adminUsers", userRepository.countByRole("ADMIN"));
        stats.put("staffUsers", userRepository.countByRole("USER"));

        return stats;
    }
}
